src/encodings/levels.rs

```diff
@@ -64,120 +97,352 @@ impl LevelEncoder {
         RleEncoder::max_buffer_size(bit_width, num_buffered_values) +
           RleEncoder::min_buffer_size(bit_width)
       },
+      Encoding::BIT_PACKED => {
+        ceil((num_buffered_values * bit_width as usize) as i64, 8) as usize
+      },
       _ => panic!("Unsupported encoding type {}", encoding)
     }
   }
 
+  /// Finalize level encoder, flush all intermediate buffers and return resulting encoded buffer.
+  /// Returned buffer is already truncated to encoded bytes only.
   #[inline]
-  pub fn consume(mut self) -> Result<Vec<u8>> {
-    self.rle_encoder.flush()?;
-    let len = (self.rle_encoder.len() as i32).to_le();
-    let len_bytes = len.as_bytes();
-    let mut encoded_data = self.rle_encoder.consume();
-    encoded_data[0..len_bytes.len()].copy_from_slice(len_bytes);
-    Ok(encoded_data)
+  pub fn consume(self) -> Result<Vec<u8>> {
+    match self {
+      LevelEncoder::RLE { bit_width: _, mut rle_encoder } => {
+        rle_encoder.flush()?;
+        let len = (rle_encoder.len() as i32).to_le();
+        let len_bytes = len.as_bytes();
+        let mut encoded_data = rle_encoder.consume();
+        encoded_data[0..len_bytes.len()].copy_from_slice(len_bytes);
+        Ok(encoded_data)
+      },
+      LevelEncoder::BIT_PACKED { bit_width: _, bit_packed_encoder } => {
+        Ok(bit_packed_encoder.consume())
+      },
+    }
   }
 }
 
-/// A encoder for definition/repetition levels. This is a thin
-/// wrapper on `RleDecoder`. Currently it only support RLE encoding.
-pub struct LevelDecoder {
-  encoding: Encoding,
-  bit_width: u8,
-  rle_decoder: RleDecoder
+/// A decoder for definition/repetition levels.
+/// Currently only supports RLE and BIT_PACKED (dev/null) encoding.
+pub enum LevelDecoder {
+  RLE { bit_width: u8, num_values: Option<usize>, rle_decoder: RleDecoder },
```

Comment:
Seems there are some duplication. Maybe we can create an enum just for the internal decoder? i.e.:
```
enum InternalDecoder {
  RLE { decoder: RleDecoder },
  BIT_PACKED { decoder: BitRaeder }
}
```

This also applies to `LevelEncoder`.